### PR TITLE
GH-37: Improve RuleExecuted event signature. GH-42: In TokenHolder.sol stored nonce (per key) should be the consumed one.

### LINKED ISSUE #37: Improve RuleExecuted event signature
The proposed signature:

```solidity
event RuleExecuted(
    address indexed _to,
    bytes4 _functionSelector,
    address _ephemeralKey,
    uint256 _nonce,
    bytes32 _messageHash,
    bool _status
);
```

### LINKED ISSUE #42: In TokenHolder.sol stored nonce (per key) should be the consumed one
The stored nonce (per key) should be the consumed nonce; so change order of adding and checking nonce; and update error message on require.

```solidity

keyData.nonce = keyData.nonce.add(1);

require(
    _nonce == keyData.nonce,
    "Nonce is not equal to the current nonce."
);
```

### PR DESCRIPTION
**Pull request by pgev**
fixes #37 
fixes #42

### CODE REVIEW

File: test/token_holder/execute_rule.js [hunk 1]
```diff
@@ -0,0 +1,605 @@
...
+        (await web3.eth.getBlockNumber()) + _deltaExpirationHeight,
+        {
+            from: registeredWallet0,
+        },
+    );
+
+    return tokenHolder;
+}
+
+contract('TokenHolder::executeRule', async () => {
```
Comment (gulshanvasnani) [inline test/token_holder/execute_rule.js]: Add test case where ephemeralkey to be authorized is passed null and empty.

File: contracts/TokenHolder.sol [hunk 2]
```diff
@@ -510,4 +527,15 @@ contract TokenHolder is MultiSigWallet {
             )
         );
     }
+
+    function bytesToBytes4(bytes _input) public pure returns (bytes4 out_) {
+        require(
+            _input.length >= 4,
+            "Input bytes length is less than 4."
```
Comment (abhayks1) [inline contracts/TokenHolder.sol]: I think it should be _input.length > 4.  We need to parse to bytes 4 if it's more than bytes4.
Also the message should be "Input bytes length is less than or equal to 4".

File: contracts/TokenHolder.sol [hunk 3]
```diff
@@ -362,13 +370,22 @@ contract TokenHolder is MultiSigWallet {
...
         // solium-disable-next-line security/no-low-level-calls
-        executeStatus_ = _to.call(_data);
+        executionStatus_ = _to.call(_data);
 
         BrandedToken(brandedToken).approve(tokenRules, 0);
 
         TokenRules(tokenRules).disallowTransfers();
 
-        emit RuleExecuted(messageHash, _nonce, executeStatus_);
+        bytes4 functionSelector = bytesToBytes4(_data);
```
Comment (abhayks1) [inline contracts/TokenHolder.sol]: We can put a require for EXECUTE_RULE_CALLPREFIX and functionSelector.

File: package.json [hunk 4]
```diff
@@ -26,6 +26,7 @@
     "eslint": "5.5.0",
     "eslint-config-airbnb-base": "13.1.0",
     "eslint-plugin-import": "2.14.0",
+    "ethereumjs-util": "^5.2.0",
```
Comment (abhayks1) [inline package.json]: "ethereumjs-util": "5.2.0" for now.

File: test/token_holder/execute_rule.js [hunk 5]
```diff
@@ -0,0 +1,605 @@
...
+    contract('Negative Testing', async (accounts) => {
+        it('Non authorized ephemeral key.', async () => {
```
Comment (abhayks1) [inline test/token_holder/execute_rule.js]: I think "it" always should make a proper statement.
e.g. it('should fail for Non authorized ephemeral key.')
For readers it's very clear.

File: test/token_holder/execute_rule.js (hunk 1 above)
Comment (pgev) [reply test/token_holder/execute_rule.js]: This is a unit test for TokenHolder::executeRule.
Are you asking to test TokenHolder::authorizeSession function here also, or?
If it's about testing TokenHolder::authorizeSession then it will be done in separate test (already implemented in PR #50).

File: test/token_holder/execute_rule.js [hunk 6]
```diff
@@ -0,0 +1,605 @@
...
+                    mockRuleSucceedActionData,
+                    invalidNonce2,
+                    rsv2.v,
+                    ethUtils.bufferToHex(rsv2.r),
+                    ethUtils.bufferToHex(rsv2.s),
+                ),
+                'Transaction is signed with non authorized key.',
+            );
+        });
+    });
```
Comment (abhayks1) [inline test/token_holder/execute_rule.js]: Few more scenarios we can add:
- execute rule failure with revoked ephemeral key
- Testing a rule with spending limit 0 success case.

File: contracts/TokenHolder.sol [hunk 7]
```diff
@@ -55,8 +55,11 @@ contract TokenHolder is MultiSigWallet {
     );
 
     event RuleExecuted(
-        bytes32 _messageHash,
+        address indexed _to,
```
[the remark this answers is not in the document]
Comment (pgev) [reply contracts/TokenHolder.sol]: The idea is to filter against specific rule. Why do you think it's not needed?

File: contracts/TokenHolder.sol (hunk 2 above)
Comment (pgev) [reply contracts/TokenHolder.sol]: We are going to convert to bytes4. Why should we check that it's bigger then 4 bytes?

File: contracts/TokenHolder.sol (hunk 3 above)
Comment (pgev) [reply contracts/TokenHolder.sol]: I am not sure I follow. Here, the functionSelector is not EXECUTE_RULE_CALLPREFIX, indeed it's a rule's function selector that should be called.

File: package.json (hunk 4 above)
Comment (pgev) [reply package.json]: Done.

File: test/token_holder/execute_rule.js (hunk 5 above)
Comment (pgev) [reply test/token_holder/execute_rule.js]: `contract` and `it` describes more than individual `assert` -s within. In some cases, when there is only one check within `it` section than we could be specific. Once we merge all unit tests I would give a bigger round and will check cases for improvements.

File: test/token_holder/execute_rule.js (hunk 6 above)
Comment (pgev) [reply test/token_holder/execute_rule.js]: 1. Added a test case for executing rule with revoked key.
2. Ticket #49 is going to handle this case.

File: contracts/TokenHolder.sol (hunk 3 above)
Comment (abhayks1) [reply contracts/TokenHolder.sol]: Sorry my bad. You are correct.

File: contracts/TokenHolder.sol [hunk 8]
```diff
@@ -453,8 +470,8 @@ contract TokenHolder is MultiSigWallet {
         EphemeralKeyData storage keyData = ephemeralKeys[key_];
 
         require(
-            _nonce == keyData.nonce,
-            "Nonce is not equal to the current nonce."
+            _nonce == keyData.nonce.add(1),
+            "The next nonce is not provided."
         );
 
         keyData.nonce = keyData.nonce.add(1);
```
[the remark this answers is not in the document]
Comment (abhayks1) [reply contracts/TokenHolder.sol]: Not 100% sure. We could avoid code repetition. Not a blocker for PR merge.

File: contracts/TokenHolder.sol [hunk 9]
```diff
@@ -352,12 +369,14 @@ contract TokenHolder is MultiSigWallet {
...
+
         require(
-            _nonce == keyData.nonce,
-            "Nonce is not equal to the current nonce."
+            _nonce == expectedNonce,
+            "The next nonce is not provided."
         );
 
-        keyData.nonce = keyData.nonce.add(1);
+        keyData.nonce = expectedNonce;
```
Comment (abhayks1) [inline contracts/TokenHolder.sol]: Approved the changes. Before merging I think we should create issue for below use case and handling it.

Scenario
- ephemeral key current nonce = 2
- Facilitator calls executeRule with higher nonce = 4(nonce 3 skipped. reason could be network delay from user wallet to blockchain) . 
- Transaction gets failed as it checks for exact nonce. Data and signature is public in blockchain for nonce=4.
- With correct nonce=3, transaction is successfully made and the gap is filled. Now expected nonce=4.
- Since the signature and data is public in blockchain for nonce=4, Should we allow 
transaction to be retried again by anyone.

Comment (pgev) [reply contracts/TokenHolder.sol]: This is a normal situation. User has signed a transaction with nonce 4 (only user can do that as an owner of private key). So, transaction with nonce 4 can be evaluated once nonce 3 was consumed.

Comment (abhayks1) [reply contracts/TokenHolder.sol]: I might be not clear : concern here is since to, data, signature is public in blockchain for nonce=4(because of failed transaction), anyone(hacker) can exploit (e.g. a big transfer) the execute rule before use calls it. I am going ahead with merge to expedite openst.js team.
